<html>
	<!-- Lovingly hand-coded in vim -->
	<head>
		<title>
			Genome-wide spatial patterns of gene expression in Drosophila
		</title>
	</head>
	<body>
		<center>
			<h3>Supplemental website for:</h3>
			<h2>
				Virtual <em>in situs</em>: Sequencing mRNA from cryo-sliced <i>Drosophila</i> embryos to determine genome-wide spatial patterns of gene expression
			</h2>
			<p>
			Peter A. Combs<sup>1</sup>, Michael B. Eisen<sup>2,3</sup>
			</p>
		</center>
		1 Graduate Program in Biophysics, University of California, Berkeley, California, United States of America,<br>
		2 Department of Molecular and Cell Biology, University of California, Berkeley, California, United States of America,<br>
		3 Howard Hughes Medical Institute, University of California, Berkeley, California, United States of America,<br>
		<br>

		<p>
		<h3>Abstract</h3></p><p>
		Complex spatial and temporal patterns of gene expression underlie embryo differentiation, yet methods do not yet exist for the efficient genome-wide determination of spatial expression patterns during development. <em>In situ</em> imaging of transcripts and proteins is the gold-standard, but it is difficult and time consuming to apply to an entire genome, even when highly automated. Sequencing, in contrast, is fast and genome-wide, but is generally applied to homogenized tissues, thereby discarding spatial information. It is likely that these methods will ultimately converge, and we will be able to sequence RNAs in situ, simultaneously determining their identity and location. As a step along this path, we developed methods to cryosection individual blastoderm stage <em>Drosophila melanogaster</em> embryos along the anterior-posterior axis and sequence the mRNA isolated from each 25Î¼m slice. The spatial patterns of gene expression we infer closely match patterns previously determined by<em> in situ</em> hybridization and microscopy. We applied this method to generate a genome-wide timecourse of spatial gene expression from shortly after fertilization through gastrulation. We identify numerous genes with spatial patterns that have not yet been described in the several ongoing systematic in situ based projects. This simple experiment demonstrates the potential for combining careful anatomical dissection with high-throughput sequencing to obtain spatially resolved gene expression on a genome-wide scale.
		</p>
		<p>
		<h3>Data</h3></p><p>
		<ul>
			<li>
			<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE43506">RNA Sequencing Reads</a>
			</li>
			<li>
			<a href="genes.cuff">Gene Abundances</a>
			</li>

		</ul>
		</p>
		<h3>Query Genes from the database</h3>
		<form action="genes_timecourse.py" method="post">
			<textarea name="genes" rows=15 cols=30>hb
gt
eve 
</textarea><br>
			<input type="submit">
		</form>
	</body>
</html>

